File: Java/Behavioral/IteratorPattern.java
package Behavioral;

import java.util.List;
import java.util.ArrayList;

interface Iterator {
    boolean hasNext();
    Object next();
}

class NameRepository {
    private List<String> names = new ArrayList<>();

    public NameRepository() {
        names.add("John");
        names.add("Jane");
        names.add("Paul");
    }

    public Iterator getIterator() {
        return new NameIterator();
    }

    private class NameIterator implements Iterator {
        int index;

        public boolean hasNext() {
            return index < names.size();
        }

        public Object next() {
            if (this.hasNext()) {
                return names.get(index++);
            }
            return null;
        }
    }
}

public class IteratorPattern {
    public static void main(String[] args) {
        NameRepository repository = new NameRepository();
        Iterator iterator = repository.getIterator();

        while (iterator.hasNext()) {
            System.out.println(iterator.next());
        }
    }
}

File: Python/Behavioral/IteratorPattern.py
class Iterator:
    def __init__(self, collection):
        self.collection = collection
        self.index = 0
    
    def has_next(self):
        return self.index < len(self.collection)

    def next(self):
        if self.has_next():
            item = self.collection[self.index]
            self.index += 1
            return item
        return None

# Usage
names = ["John", "Jane", "Paul"]
iterator = Iterator(names)

while iterator.has_next():
    print(iterator.next())  # Output: John, Jane, Paul
